django/db/models/indexes.py

```diff
@@ -31,26 +32,44 @@ def __init__(
             raise ValueError('Index.fields must be a list or tuple.')
         if not isinstance(opclasses, (list, tuple)):
             raise ValueError('Index.opclasses must be a list or tuple.')
-        if opclasses and len(fields) != len(opclasses):
-            raise ValueError('Index.fields and Index.opclasses must have the same number of elements.')
-        if not fields:
-            raise ValueError('At least one field is required to define an index.')
+        if not expressions and not fields:
+            raise ValueError('At least one field or expression is required to define an index.')
+        if fields and not all(isinstance(field, str) for field in fields):
+            raise ValueError('Index.fields must only contain strings.')
+        if expressions and fields:
+            raise ValueError("'fields' cannot be used together with expressions.")
+        if expressions and not name:
+            raise ValueError('Index.name needs to be set when passed expressions.')
+        if opclasses:
+            if expressions and len(expressions) != len(opclasses):
+                raise ValueError('Index arguments and Index.opclasses must have the same number of elements.')
+            elif fields and len(fields) != len(opclasses):
+                raise ValueError('Index.fields and Index.opclasses must have the same number of elements.')
         if include and not name:
             raise ValueError('A covering index must be named.')
         if not isinstance(include, (type(None), list, tuple)):
             raise ValueError('Index.include must be a list or tuple.')
         self.fields = list(fields)
-        # A list of 2-tuple with the field name and ordering ('' or 'DESC').
-        self.fields_orders = [
-            (field_name[1:], 'DESC') if field_name.startswith('-') else (field_name, '')
-            for field_name in self.fields
-        ]
         self.name = name or ''
         self.db_tablespace = db_tablespace
         self.opclasses = opclasses
         self.condition = condition
         self.include = tuple(include) if include else ()
 
+        self.field_names = [
+            field_name.lstrip('-')
+            for field_name in self.fields
+        ]
+        # TODO: Do we need to keep original args for `deconstruct()` or can we convert strings to `F()`?
+        self.expressions = tuple(
+            F(expression) if isinstance(expression, str) else expression
+            for expression in expressions
+        )
```

Comment:
As long as they both equals it shouldn't make a difference.

By that I mean you should have a test asserts `Index('foo') == Index(F('foo'))`

An alternative could be to defer that to `create_sql`.

Reply:
I’ll add a test for it in `tests.model_indexes.tests.SimpleIndexesTests`, looks like I’ve missed that.